### PR TITLE
rust-analyzer: 2020-08-24 -> 2020-09-14 and some clean-up

### PR DESCRIPTION
**Pull request by oxalica**
###### Motivation for this change
Bump version and do some clean-up.

- `unstable-2020-09-07` requires rustc 1.46.0 to compile but we still have 1.45.2 in master. Patches are introduced for a temporary fix.
- Feature `jemalloc` is removed now, and a new features `mimalloc` is introduced.
- Remove fake rustup in check phase, since we currently test only sub-crate `rust-analyzer`, which doesn't need it anymore.
- Install check works now. Enable it by default.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

r? @danieldk

### CODE REVIEW

**Comment by danieldk** [comment]
This update touches node package, since I literally know nothing about node, I'll leave up the reviewing to someone else.

File: pkgs/development/tools/rust/rust-analyzer/generic.nix
Comment (midchildan) [inline pkgs/development/tools/rust/rust-analyzer/generic.nix:21]: Note: Upstream PR in https://github.com/rust-analyzer/rust-analyzer/pull/5354

File: pkgs/development/tools/rust/rust-analyzer/generic.nix
Comment (midchildan) [inline pkgs/development/tools/rust/rust-analyzer/generic.nix:23]: Note: Needed for `mimalloc_rust` to build its bundled version of the mimalloc library. https://github.com/purpleprotocol/mimalloc_rust#requirements

File: pkgs/development/tools/rust/rust-analyzer/generic.nix
Comment (midchildan) [inline pkgs/development/tools/rust/rust-analyzer/generic.nix:2]: Is it necessary to split the arguments? Looking at the output of `rg -B3 '@args'`, it doesn't look like a common pattern in nixpkgs.

File: pkgs/development/tools/rust/rust-analyzer/default.nix
Comment (midchildan) [inline pkgs/development/tools/rust/rust-analyzer/default.nix:14]: I think this would better belong in `generic.nix`, along with other build instructions.

**Review by midchildan — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/rust/rust-analyzer/default.nix
Comment (oxalica) [reply pkgs/development/tools/rust/rust-analyzer/default.nix:14]: :thinking: But it's temporary.

Well, I'm not quite sure what should be put in generic file. You mean all about "how to build", yeah?

Comment (midchildan) [reply pkgs/development/tools/rust/rust-analyzer/default.nix:14]: I'd expect most patches are supposed to be temporary workarounds, so it wouldn't be a problem as long as there's a comment explaining why they're necessary. If you want to make absolutely sure that you remove the patch when it's no longer needed, I guess you could do something like this:

```nix
  patches = with lib;
    if versionAtLeast (getVersion rustPlatform.rust.rustc) "1.46.0" then [
      # FIXME: Temporary fix for our rust 1.45.0 since rust-analyzer requires 1.46.0
      ./no-loop-in-const-fn.patch
      ./no-option-zip.patch
    ] else warn "Some patches for ${pname} needs to be removed." [];
```

This would exclude the patches and warn you about it once `rustc` reaches 1.46.0.

> You mean all about "how to build", yeah?

Yes.

Comment (oxalica) [reply pkgs/development/tools/rust/rust-analyzer/default.nix:14]: > If you want to make absolutely sure that you remove the patch when it's no longer needed, I guess you could do something like this

These patches just do the same thing in a more compatible way and don't change any semantics. I think it's not quite necessary to do that check.

File: pkgs/development/tools/rust/rust-analyzer/generic.nix
Comment (midchildan) [inline pkgs/development/tools/rust/rust-analyzer/generic.nix:7]: Could you also make this a simple function instead of a nested one?

**Review by midchildan — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/rust/rust-analyzer/generic.nix
Comment (oxalica) [reply pkgs/development/tools/rust/rust-analyzer/generic.nix:7]: Is it necessary? Some are just extracted version related parameters, while others are passed through `callPackages` which can be overrided. It's not expected to directly configure rev, version and hashes like other packages in nixpkgs.

Comment (midchildan) [reply pkgs/development/tools/rust/rust-analyzer/generic.nix:7]: It's only a matter of style but I found the original version easier to read than the nested version, although they basically do the same thing.

original:
```nix
rust-analyzer-unwrapped = callPackage ./generic.nix rec {
  rev = ...;
  ...
};
```

nested:
```nix
rust-analyzer-unwrapped = callPackage (import ./generic.nix rec {
  rev = ...;
  ...
}) {};
```

**Review by midchildan — APPROVED** [review]
LGTM. The remaining point is a minor style issue, so I don't see much point holding this up.

**Comment by midchildan** [comment]
@oxalica This still needs to go through committer review, but you might want to squash the commits in the meantime.

File: pkgs/development/tools/rust/rust-analyzer/default.nix
Comment (midchildan) [reply pkgs/development/tools/rust/rust-analyzer/default.nix]: That makes sense.

**Comment by oxalica** [comment]
The global node-packages (`pkgs/development/node-packages/node-packages.nix`) is updated too frequently.
I'm thinking about using a standalone generated `node-packages.nix` for the package.

What's the best practice about these npm packages as only build-dependency?

@jonringer

**Comment by oxalica** [comment]
I refactored the node packages part to use generated standalone `node-packages.nix` to avoid touching global `nodePackages`.

@midchildan Have a look? What do you think about it?

**Comment by midchildan** [comment]
Using node2nix outside of `nodePackages` is generally frowned upon [[1],[2]] even if it's only used as a build dependency [[3],[4]]. The guideline for node.js packaging [[5]] states that "the package set should only provide end user software package," and this has been pointed out in another issue [[6]], but I think it's best to stick with the currently accepted practice for this specific PR.

[1]: https://github.com/NixOS/nixpkgs/pull/78810
[2]: https://github.com/NixOS/nixpkgs/pull/49413
[3]: https://github.com/NixOS/nixpkgs/pull/68820
[4]: https://github.com/NixOS/nixpkgs/pull/69219
[5]: https://github.com/NixOS/nixpkgs/blob/2b575c5f1f287293a5dbfab3e9963519e713fd3b/doc/languages-frameworks/node.section.md
[6]: https://github.com/NixOS/nixpkgs/issues/82662

**Comment by oxalica** [comment]
@midchildan I'm just tired of generating and generating again over **ALL** NPM packages used by nixpkgs **every time** having merge conflict (it's updated quite often). We have `package-lock.json` but `node2nix` ignores them and fetches all packages every time, which takes about an hour for me and (accidentally) cause ~100 package rebuild (which may be not needed or even lead to build failure)

(Also `node2nix` is poor designed and the fetching is synchronous. It takes years in a proxy environment with high latency but large bandwidth.)

Note that just when I typing this comment, there's another update on `pkgs/development/node-packages/node-packages.nix`. 
MERGE CONFLICT AGAIN :disappointed:

**Comment by midchildan** [comment]
The current situation isn't ideal, but I believe this discussion would be best done in a separate issue because this affects how node packaging should be done in general, and isn't limited to this package. #82662 seems like a good place to start.

For now, I'd recommend ignoring merge conflicts until you receive committer approval. You can perform a rebase afterwards.

**Review by midchildan — APPROVED** [review]

**Review by midchildan — APPROVED** [review]

**Review by jonringer — APPROVED** [review]
LGTM

failures are broken on target branch
```
https://github.com/NixOS/nixpkgs/pull/97444
2 packages failed to build:
dat vimPlugins.coc-imselect

17 packages built:
bitwarden-cli castnow epgstation joplin lumo mirakurun netlify-cli redoc-cli rust-analyzer rust-analyzer-unwrapped thelounge vimPlugins.coc-eslint vimPlugins.coc-git vimPlugins.coc-metals vimPlugins.coc-prettier vimPlugins.coc-stylelint vimPlugins.coc-vetu
```

**Pull request merged**
